src/core/meshcop/joiner.cpp

```diff
@@ -197,19 +201,32 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
         steeringData.SetLength(aResult->mSteeringData.mLength);
         memcpy(steeringData.GetValue(), aResult->mSteeringData.m8, steeringData.GetLength());
 
-        if (steeringData.DoesAllowAny() ||
-            (steeringData.GetBit(mCcitt % steeringData.GetNumBits()) &&
-             steeringData.GetBit(mAnsi % steeringData.GetNumBits())))
+        if (steeringData.DoesAllowAny())
         {
-            mJoinerUdpPort = aResult->mJoinerUdpPort;
-            mJoinerRouterPanId = aResult->mPanId;
-            mJoinerRouterChannel = aResult->mChannel;
-            memcpy(&mJoinerRouter, &aResult->mExtAddress, sizeof(mJoinerRouter));
+            VerifyOrExit(!mJoinerRouterIsSpecific);
+            VerifyOrExit(aResult->mRssi > mJoinerRouterRssi);
+        }
+        else if (steeringData.GetBit(mCcitt % steeringData.GetNumBits()) &&
+                 steeringData.GetBit(mAnsi % steeringData.GetNumBits()))
+        {
+            if (mJoinerRouterIsSpecific)
+            {
+                VerifyOrExit(aResult->mRssi > mJoinerRouterRssi);
+            }
+
+            mJoinerRouterIsSpecific = true;
         }
         else
         {
             otLogDebgMeshCoP(GetInstance(), "Steering data does not include this device");
+            ExitNow();
         }
+
+        mJoinerUdpPort = aResult->mJoinerUdpPort;
+        mJoinerRouterPanId = aResult->mPanId;
+        mJoinerRouterChannel = aResult->mChannel;
+        mJoinerRouterRssi = aResult->mRssi;
```

Comment:
Thanks for making the updates.  FYI, Travis is showing a [compiler warning](https://travis-ci.org/openthread/openthread/jobs/269661927) on this line.

Reply:
My punishment for committing right at the end of the working day, of course! Fixed, and i'll keep an eye out for anything that pops up this build